File: kwonmin/23_09/programmers/64062.java
// 생각 전개부터 하자
// 몇 명 까지 가능한지
// 범위 : 200000000 이네요
// 이분탐색으로 가능할듯?
// 2억이니껜,,, 헉
// 처음에 max를 잡을지 2만을 잡을 지 고민.
// max = n이 드니껜. 그래도 최초 범위를 줄이는 게 더 낫지 않을까?


public class Niniz {

	public int solution(int[] stones, int k) {

		int start = 1;
		int end = 200000000;

		while(start <= end){
			int mid = (start+end)/2;
			boolean flag = false;
			int empty = 0;

			for (int i = 0; i < stones.length; i++) {
				int stone = stones[i] - mid;
				if (stone <= 0) {
					empty++;
					if (empty >= k) {
						flag = true;
						break;
					}
				} else {
					empty = 0;
				}
			}
			if(flag) end = mid-1;
			else start = mid+1;

		}


		return start-1;
	}
}

File: eunjin/23_09/64062.py
def solution(stones, k):
    answer = 0
    left, right = 0, max(stones)
    while left <= right:
        mid = (left + right) // 2
        j = 0
        jump = 0
        while j <= len(stones):
            if stones[j] - mid <= 0:
                jump += 1
            else:
                jump = 0
            if jump >= k:
                break
            j += 1
        if jump < k:
            left = mid + 1
        else:
            answer = mid
            right = mid - 1
    return answer

print(solution([2, 4, 5, 3, 2, 1, 4, 2, 5, 1], 3))